File: tests/src/org.joda/joda-convert/2.2.4/src/test/java/org_joda/joda_convert/TypeUtilsTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_joda.joda_convert;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Map;

import org.joda.convert.StringConvert;
import org.junit.jupiter.api.Test;

public class TypeUtilsTest {
    @Test
    public void parsesSourceAndJvmArraySyntaxInParameterizedTypes() {
        StringConvert convert = StringConvert.create();

        ParameterizedType type = convert.convertFromString(
                ParameterizedType.class,
                "java.util.Map<java.lang.String[], [Ljava.lang.Integer;>");
        Type[] arguments = type.getActualTypeArguments();

        assertEquals(Map.class, type.getRawType());
        assertArrayEquals(new Type[] {String[].class, Integer[].class}, arguments);
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/TypeUtilsTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

import org.joda.convert.StringConvert
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class TypeUtilsTest {
  @Test
  def parsesJavaSourceStyleArrayTypeArguments(): Unit = {
    val parsed: Type = new StringConvert().convertFromString(
      classOf[Type],
      "java.util.List<java.lang.String[]>"
    )

    val parameterized: ParameterizedType = asParameterizedType(parsed)
    assertEquals(classOf[java.util.List[?]], parameterized.getRawType)
    assertEquals(classOf[Array[String]], parameterized.getActualTypeArguments.apply(0))
  }

  @Test
  def parsesJvmDescriptorStyleArrayTypeArguments(): Unit = {
    val parsed: Type = new StringConvert().convertFromString(
      classOf[Type],
      "java.util.List<[Ljava.lang.String;>"
    )

    val parameterized: ParameterizedType = asParameterizedType(parsed)
    assertEquals(classOf[java.util.List[?]], parameterized.getRawType)
    assertEquals(classOf[Array[String]], parameterized.getActualTypeArguments.apply(0))
  }

  private def asParameterizedType(parsed: Type): ParameterizedType = {
    assertTrue(
      parsed.isInstanceOf[ParameterizedType],
      s"Expected a parameterized type but got ${parsed.getTypeName}"
    )
    parsed.asInstanceOf[ParameterizedType]
  }
}
